nixio/cmd/nixexplore.py

```diff
@@ -0,0 +1,623 @@
+    def name_lambda(n, full_match):
+        if full_match:
+            return lambda s: n.lower() == s.name.lower()
+        else:
+            return lambda s: n.lower() in s.name.lower()
+
+    def name_lambda_cs(n, full_match):
+        if full_match:
+            return lambda s: n == s.name
+        else:
+            return lambda s: n in s.name
+
+    secs = nix_file.find_sections(type_lambda(pattern, full_match))
+    if len(secs) == 0:
+        secs = nix_file.find_sections(name_lambda_cs(pattern, full_match) if case_sensitive else name_lambda(pattern, full_match))
+    return secs
+
+
+def find_props(nix_file, pattern, case_sensitive=False, full_match=False):
+    n = pattern if case_sensitive else pattern.lower()
+    props = {}
+    for s in nix_file.find_sections(lambda s: True):
+        for p in s.props:
+            pname = p.name if case_sensitive else p.name.lower()
+            if n == pname if full_match else n in pname:
+                if s not in props.keys():
+                    props[s] = []
+                props[s].append(p)
+    return props
+
+
+def disp_metadata(filename, arguments):
+    case_sensitive = arguments.case_sensitive
+    full_match = arguments.full_match
+    f = open_nix_file(filename)
+    if f is None:
+        pass
+    print("%s: " % (filename.split(os.sep)[-1]), end="\n")
+    if arguments.pattern is None or len(arguments.pattern) == 0:
+        for sec in f.sections:
+            sec.pprint(max_depth=arguments.depth)
+    else:
+        for p in arguments.pattern:
+            if "/" in p:
+                parts = p.split("/")
+                secs = find_section(f, parts[0], case_sensitive)
+                for s in secs:
+                    for prop in s.props:
+                        part = parts[1] if case_sensitive else parts[1].lower()
+                        pname = prop.name if case_sensitive else prop.name.lower()
+                        if part == pname if full_match else part in pname:
+                            print("[section: %s, type: %s, id: %s] >> " % (s.name, s.type, s.id), end="")
+                            prop.pprint()
+            else:
+                secs = find_section(f, p, case_sensitive)
+                if len(secs) == 0:
+                    props = find_props(f, p, case_sensitive)
+                    for sec in props.keys():
+                        for prop in props[sec]:
+                            print("[section: %s, type: %s, id: %s] >> " % (sec.name, sec.type, sec.id), end="")
+                            prop.pprint()
+                else:
+                    for s in secs:
+                        s.pprint(max_depth=arguments.depth)
+
+    f.close()
+    print("\n\n")
+
+
+def mdata_worker(arguments):
+    files = assemble_files(arguments)
+    for nf in files:
+        disp_metadata(nf, arguments)
+
+
+def find_data_entity(nix_file, arguments):
+    name_or_type = arguments.pattern if arguments.case_sensitive else arguments.pattern.lower()
```

Comment:
Causes error if no pattern is supplied on the command line.  Should patterns be required or should this pretty-print the data tree when no patterns are supplied.  I'd go for printing the whole data tree for consistency with the `mdata` subcommand.

Reply:
actually, pattern expects one argument, if no argument is given, there will be an error raised by argparse. If you pass an empty string, it will print out a summary of all stored entities.

Reply:
I mean that the `-p` flag is optional but this line assumes it's not and tries to read it.  So if you run something like `nixio-explore dump <filename>` you get
```
Traceback (most recent call last):
File "/home/achilleas/.local/bin/nixio-explore", line 33, in <module>
sys.exit(load_entry_point('nixio', 'console_scripts', 'nixio-explore')())
File "/home/achilleas/projects/nixpy/nixio/cmd/nixexplore.py", line 622, in main
args.func(args)
File "/home/achilleas/projects/nixpy/nixio/cmd/nixexplore.py", line 523, in dump_worker
func(nf, arguments)
File "/home/achilleas/projects/nixpy/nixio/cmd/nixexplore.py", line 468, in data_dump
entities = find_data_entity(nix_file, arguments)
File "/home/achilleas/projects/nixpy/nixio/cmd/nixexplore.py", line 332, in find_data_entity
name_or_type = arguments.pattern if arguments.case_sensitive else arguments.pattern.lower()
AttributeError: 'NoneType' object has no attribute 'lower'
```
because `-p` wasn't supplied and `arguments.pattern` is None.  We should either check for None or set the `-p` flag to have a default value of empty string.

Reply:
ah I see, was testing with the data subcommand

Reply:
should be solved, pattern got defaults now